### DISCUSSION TITLE
Scaling billboard text with a transform?

### DISCUSSION THREAD

**Discussion by gwaldron** [vsg-dev/VulkanSceneGraph#985] [Q&A] [answered]
Hi Robert,
In #619 you recommend scaling a `vsg::Text` node with a `MatrixTransform`, but I can't get this to work properly when billboarding is enabled. I am using vsg 1.0.9.

To reproduce this, you can alter vsgtext.cpp to insert a scaling transform on one of the nodes:

(Line 287):
```c++
#if 0
            scenegraph->addChild(text);
#else
            auto mt = vsg::MatrixTransform::create();
            mt->matrix = vsg::scale(5.0);
            mt->addChild(text);
            scenegraph->addChild(mt);
#endif
```

Original:
<img width="653" alt="image" src="https://github.com/vsg-dev/VulkanSceneGraph/assets/326618/e265c262-b7e3-4565-8273-75e7b2815099">

With the scale transform:
<img width="475" alt="image" src="https://github.com/vsg-dev/VulkanSceneGraph/assets/326618/02b9e43b-462d-492d-b5ff-a019495cee0a">

It scales the position itself, which is interesting, but the actual size of the text doesn't change.

Any advice on this?

**Comment by robertosfield** [comment]
Could it be that the default camera position is simply moving back to encompass the now larger scene?

**Reply by gwaldron** [reply]
Well, yes it is, but the text still isn't scaling.

**Reply by robertosfield** [reply]
Is the text in question set up to automatically scaling to screen size?

Also, the StandardLayout has horizontal and vertical size parameter is there a reason why you aren't using this?

**Reply by gwaldron** [reply]
Because I want to scale dynamically, and in #619 you said:

> You can also decorate a vsg::Text/TextGroup with a MatrixTransform to scale them, if you want to vary the size dynamically then that's the way I'd do it.

I don't know if it's set up to scale to screen size. Right now I'm just running the vsgtext example and the setup looks like this:
```c++
auto layout = vsg::StandardLayout::create();
layout->horizontalAlignment = vsg::StandardLayout::CENTER_ALIGNMENT;
layout->position = vsg::vec3(0.0, 0.0, 0.0);
layout->horizontal = vsg::vec3(1.0, 0.0, 0.0);
layout->vertical = vsg::vec3(0.0, 1.0, 0.0);
layout->color = vsg::vec4(1.0, 1.0, 1.0, 1.0);
layout->outlineWidth = 0.1;
layout->billboard = true;
```

**Reply by robertosfield** [reply]
Billboarded text rebuilds the modelview matrix within the vertex shader and ignores the previous scale, so if you are attempting to scale billboarded text then it won't work. 

https://github.com/vsg-dev/vsgExamples/blob/master/data/shaders/standard.vert#L36

Use of the MatrixTransform to scale a subgraph will work when the text is standard billboard text.

What exactly is the end result you are after?

**Reply by gwaldron** [reply]
I want to render a text label and size it in points or screen pixels, regardless of its position in the scene. Think labeling an object on the globe.

I've solved this another way for now, with my own custom transform about the vsg::Text and setting the layout->billboard to false.

https://github.com/pelicanmapping/rocky/blob/main/src/rocky_vsg/PixelScaleTransform.h

<img width="343" alt="image" src="https://github.com/vsg-dev/VulkanSceneGraph/assets/326618/cc616d2c-63c5-4d25-b2bb-c12ea7b77119">

**Reply by robertosfield** [reply, marked as answer]
vsg::Text is able to do this out of the box, you should just need to set the vsg::StandardLayout with billboard set to true, set the screen size via the horizontal and vertical parameters, and set the billboardAutoScaleDistance to the maximum distance that you want the text to be scaled until it is simply scaled as if object coordinates.

I implemented this combination to enable one to control the screensize when close enough but then have it shrink as other object space objects do when you are far enough away that text labels would otherwise overlap.

This scaling is done entirely by the vertex shader in the code I linked to above, so it's very lightweight and scales nicely for large numbers of text labels.
